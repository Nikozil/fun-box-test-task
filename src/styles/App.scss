@import url('https://fonts.googleapis.com/css2?family=Nunito&family=Roboto&display=swap');

.app {
  display: grid;
  grid-template-areas:
    'head head'
    'interface map';
  grid-template-rows: 60px 1fr;
  grid-template-columns: 300px 1fr;

  min-height: 100vh;
  min-width: 350px;

  font-family: 'Nunito', sans-serif;

  @media screen and (max-width: 767px) {
    grid-template-columns: 200px 1fr;
  }
}
.interface {
  grid-area: interface;
  background-color: #fff;
  padding: 10px;
}

.map {
  grid-area: map;
  background-color: #f7f7f7;
}
