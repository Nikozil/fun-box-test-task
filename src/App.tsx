import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import DotsListComponent from './components/DotsListComponent';
import Header from './components/Header';
import MapComponent from './components/MapComponent';
import NewDotForm from './components/NewDotForm';
import { reorder } from './lib/Functions';
import './styles/App.scss';

function App() {
  const [center, setCenter] = useState<Coordinates>([55.734108, 37.623915]);
  const [dotsList, setDotsList] = useState<Dot[]>([]);

  const handleSubmit = (name: DotName) => {
    const newDot = {
      id: uuidv4(),
      name: name,
      coordinates: center,
    };

    setDotsList((prevlist) => [...prevlist, newDot]);
  };

  const deleteDot = (id: DotId) => {
    setDotsList((prevlist) => prevlist.filter((dot: Dot) => dot.id !== id));
  };

  const permutationDots = (startIndex: number, endIndex: number) => {
    const newList = reorder(dotsList, startIndex, endIndex);
    setDotsList(newList);
  };

  return (
    <div className="app">
      <Header />
      <div className="interface">
        <NewDotForm handleSubmit={handleSubmit} />
        <DotsListComponent
          dotsList={dotsList}
          deleteDot={deleteDot}
          permutationDots={permutationDots}
        />
      </div>
      <div className="map">
        <MapComponent
          center={center}
          setCenter={setCenter}
          dotsList={dotsList}
          setDotsList={setDotsList}
        />
      </div>
    </div>
  );
}

export default App;

export interface Dot {
  id: DotId;
  name: DotName;
  coordinates: Coordinates;
}
export type DotId = string;
type DotName = string;
export type Coordinates = number[];
