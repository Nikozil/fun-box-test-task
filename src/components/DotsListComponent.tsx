import React from 'react';
import {
  DragDropContext,
  Draggable,
  Droppable,
  DropResult,
} from 'react-beautiful-dnd';
import { Dot, DotId } from '../App';
import styles from '../styles/DotsListComponent.module.scss';
import DotComponent from './DotComponent';

const DotsListComponent: React.FC<Props> = ({
  dotsList,
  deleteDot,
  permutationDots,
}) => {
  const onDragEnd = (result: DropResult) => {
    if (!result.destination) {
      return;
    }
    permutationDots(result.source.index, result.destination.index);
  };

  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <Droppable droppableId="droppable">
        {(provided, snapshot) => (
          <ul
            {...provided.droppableProps}
            ref={provided.innerRef}
            className={styles['dots-list__ul']}>
            {dotsList.map((dot, i) => (
              <Draggable key={dot.id} draggableId={dot.id} index={i}>
                {(provided) => (
                  <li
                    ref={provided.innerRef}
                    {...provided.draggableProps}
                    {...provided.dragHandleProps}
                    className={styles['dots-list__li']}>
                    <DotComponent dot={dot} deleteDot={deleteDot} />
                  </li>
                )}
              </Draggable>
            ))}
            {provided.placeholder}
          </ul>
        )}
      </Droppable>
    </DragDropContext>
  );
};

export default DotsListComponent;

interface Props {
  dotsList: Dot[];
  deleteDot: (id: DotId) => void;
  permutationDots: (startIndex: number, endIndex: number) => void;
}
