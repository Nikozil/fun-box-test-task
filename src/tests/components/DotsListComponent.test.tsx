import { render } from '@testing-library/react';
import React from 'react';
import { Dot } from '../../App';
import DotsListComponent from '../../components/DotsListComponent';

const dotsList = [
  { name: 'Dot 1' },
  { name: 'Dot 2' },
  { name: 'Dot 3' },
] as Dot[];
const deleteDot = jest.fn();
const permutationDots = jest.fn();

describe('DotsListComponent tests', () => {
  test('renders correct', () => {
    const { asFragment } = render(
      <DotsListComponent
        dotsList={dotsList}
        deleteDot={deleteDot}
        permutationDots={permutationDots}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
