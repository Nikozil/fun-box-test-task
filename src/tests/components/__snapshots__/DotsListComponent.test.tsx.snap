// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotsListComponent tests renders correct 1`] = `
<DocumentFragment>
  <ul
    class="dots-list__ul"
    data-rbd-droppable-context-id="0"
    data-rbd-droppable-id="droppable"
  >
    <li
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="dots-list__li"
      data-rbd-drag-handle-context-id="0"
      data-rbd-draggable-context-id="0"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div
        class="dot"
      >
        Dot 1 
        <button
          class="dot__delete-button"
          data-testid="delete-button"
        >
          <img
            alt="Delete"
            src="trash-alt-solid.svg"
          />
        </button>
      </div>
    </li>
    <li
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="dots-list__li"
      data-rbd-drag-handle-context-id="0"
      data-rbd-draggable-context-id="0"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div
        class="dot"
      >
        Dot 2 
        <button
          class="dot__delete-button"
          data-testid="delete-button"
        >
          <img
            alt="Delete"
            src="trash-alt-solid.svg"
          />
        </button>
      </div>
    </li>
    <li
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="dots-list__li"
      data-rbd-drag-handle-context-id="0"
      data-rbd-draggable-context-id="0"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <div
        class="dot"
      >
        Dot 3 
        <button
          class="dot__delete-button"
          data-testid="delete-button"
        >
          <img
            alt="Delete"
            src="trash-alt-solid.svg"
          />
        </button>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
